src/pcapi/core/bookings/api.py

```diff
@@ -82,6 +83,14 @@ def book_offer(
 
         stock.dnBookedQuantity += booking.quantity
 
+        if feature_queries.is_active(FeatureToggle.ENABLE_ACTIVATION_CODES) and stock.activationCodes:
+            validation.check_has_available_activation_code(stock.activationCodes)
+            booking.activationCode = next(
+                activationCode
+                for activationCode in stock.activationCodes
+                if check_activation_is_bookable(activationCode)
+            )
```

Comment:
J'ai l'impression qu'on fait 2 fois la même vérification (plus ou moins). Suggestion :

Ici :

```suggestion
            code = get_available_activation_code(stock)
            if code is None:
                # On ne devrait pas arriver ici, normalement. On a déjà vérifié plus haut que le `Stock.quantity`
                # permettait de faire une réservation. Donc, normalement, il devrait rester au moins un code
                # d'activation disponible.
                raise NoActivationCodeAvailable()
            booking.activationCode = code
```

Et dans `bookings/api.py` :

```python

def get_available_activation_code(stock: Stock):
    for code in stock.activationCodes:
        if code.bookingId:
            continue
        if activation_code.expirationDate and activation_code.expirationDate > datetime.datetime.utcnow():
            continue
        return code
    return None
```